core/lib/src/shutdown.rs

```diff
@@ -0,0 +1,92 @@
+use futures::{channel::mpsc, stream::StreamExt};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Wrapper around an mpsc channel.
+#[derive(Debug)]
+crate struct Shutdown {
+    sender: mpsc::Sender<()>,
+    crate receiver: Option<Receiver>,
+}
+
+impl Shutdown {
+    /// Create a `Shutdown`.
+    #[inline]
+    crate fn new() -> Self {
+        let (sender, receiver) = mpsc::channel(1);
+        Self {
+            sender,
+            receiver: Some(Receiver { receiver }),
+        }
+    }
+
+    /// Get a handle that can be used to call `.shutdown()`.
+    #[inline]
+    crate fn get_handle(&self) -> ShutdownHandle<'_> {
+        ShutdownHandle {
+            sender: &self.sender,
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # #![feature(proc_macro_hygiene, async_await)]
+/// # #[macro_use] extern crate rocket;
+/// #
+/// use rocket::shutdown::ShutdownHandle;
+///
+/// #[get("/shutdown")]
+/// fn shutdown(handle: &ShutdownHandle) -> &'static str {
+///     handle.shutdown();
+///     "Shutting down..."
+/// }
+///
+/// fn main() {
+///     # if false {
+///     rocket::ignite()
+///         .mount("/", routes![shutdown])
+///         .launch();
+///     # }
+/// }
+/// ```
+#[derive(Debug, Clone, Copy)]
+pub struct ShutdownHandle<'a> {
```

Comment:
I can't guess how likely this will be, but I will not be surprised if someone asks for a `ShutdownHandle` that is `'static`.

Reply:
How should this be reconciled with the migration to managed state?

Reply:
Also, the reason I did this was to avoid unnecessary clones. It wouldn't be much cost, but all I'd have to do is revert a single commit.